### Links to dataset
Dataset link 1: https://data.telangana.gov.in/dataset/tourism-foreign-visitors-data
<br>
Dataset link 2: https://data.telangana.gov.in/dataset/tourism-domestic-visitors-data

### Data Merging

In [1]:
import pandas as pd

# List of CSV files to combine
file_paths = ['Tourism Foreign Visitors Data 2023.csv', 'Tourism Foreign Visitors Data 2022.csv', 'Tourism Foreign Visitors Data 2021.csv', 
    'Tourism Foreign Visitors Data 2020.csv', 'Tourism Foreign Visitors Data 2019.csv', 'Tourism Foreign Visitors Data 2018.csv',
    'Tourism Foreign Visitors Data 2017.csv', 'Tourism Foreign Visitors Data 2016.csv', 'Tourism Foreign Visitors Data 2015.csv',
    'Tourism Foreign Visitors Data 2014.csv', 'Tourism Domestic Visitors Data 2023.csv', 'Tourism Domestic Visitors Data 2022.csv',
    'Tourism Domestic Visitors Data 2021.csv', 'Tourism Domestic Visitors Data 2020.csv', 'Tourism Domestic Visitors Data 2019.csv',
    'Tourism Domestic Visitors Data 2018.csv', 'Tourism Domestic Visitors Data 2017.csv', 'Tourism Domestic Visitors Data 2016.csv',
    'Tourism Domestic Visitors Data 2015.csv', 'Tourism Domestic Visitors Data 2014.csv']

# List to hold the DataFrames
dfs = []

for file_path in file_paths:
    df = pd.read_csv(file_path)
    dfs.append(df)

# Concatenate all DataFrames
combined_df = pd.concat(dfs, ignore_index=False)
combined_df.to_csv("Combined_Tourism_Visitor's_data.csv", index=True)
print('Files combined successfully!') 

Files combined successfully!


In [2]:
df = df.copy() 

In [3]:
df = pd.read_csv('Tourism Foreign Visitors Data 2023.csv')
df['Year'] = 2023
output_path = 'Tourism Foreign Visitors Data 2023.csv'
df.to_csv(output_path, index=False)
print(df.head())

   District     Month  Visitors  Year
0  Adilabad   January         0  2023
1  Adilabad  February         0  2023
2  Adilabad     March         0  2023
3  Adilabad     April         0  2023
4  Adilabad       May         0  2023


### Data Extraction

In [4]:
import numpy as np

In [5]:
df = pd.read_csv("Combined_Tourism_Visitor's_data.csv")
df.head(3)

Unnamed: 0  District     Month Visitors  Year  Unnamed: 3  Unnamed: 4  \
0           0  Adilabad   January        0  2023         NaN         NaN   
1           1  Adilabad  February        0  2023         NaN         NaN   
2           2  Adilabad     March        0  2023         NaN         NaN   

   Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9  Unnamed: 10  \
0         NaN         NaN         NaN         NaN         NaN          NaN   
1         NaN         NaN         NaN         NaN         NaN          NaN   
2         NaN         NaN         NaN         NaN         NaN          NaN   

   Unnamed: 11  Unnamed: 12  Unnamed: 13  
0          NaN          NaN          NaN  
1          NaN          NaN          NaN  
2          NaN          NaN          NaN

In [6]:
df.tail(3)

Unnamed: 0         District     Month Visitors  Year  Unnamed: 3  \
6667         369  Yadadri Bhongir   October   584800  2014         NaN   
6668         370  Yadadri Bhongir  November   681000  2014         NaN   
6669         371  Yadadri Bhongir  December   462800  2014         NaN   

      Unnamed: 4  Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9  \
6667         NaN         NaN         NaN         NaN         NaN         NaN   
6668         NaN         NaN         NaN         NaN         NaN         NaN   
6669         NaN         NaN         NaN         NaN         NaN         NaN   

      Unnamed: 10  Unnamed: 11  Unnamed: 12  Unnamed: 13  
6667          NaN          NaN          NaN          NaN  
6668          NaN          NaN          NaN          NaN  
6669          NaN          NaN          NaN          NaN

### Metadata:
1. Unnamed: 0: Visitor's Identification number
2. District : Name of the district
3. Month : Name of the month
4. Visitors : Number of visitors
5. Year: Visiting year

### Data Exploration and Manipulation

In [7]:
df.columns.to_list()

['Unnamed: 0',
 'District',
 'Month',
 'Visitors',
 'Year',
 'Unnamed: 3',
 'Unnamed: 4',
 'Unnamed: 5',
 'Unnamed: 6',
 'Unnamed: 7',
 'Unnamed: 8',
 'Unnamed: 9',
 'Unnamed: 10',
 'Unnamed: 11',
 'Unnamed: 12',
 'Unnamed: 13']

In [8]:
df.shape

(6670, 16)

 ##### There are 6670 rows and 16 columns in the dataset.

In [9]:
df.dtypes

Unnamed: 0       int64
District        object
Month           object
Visitors        object
Year             int64
Unnamed: 3     float64
Unnamed: 4     float64
Unnamed: 5     float64
Unnamed: 6     float64
Unnamed: 7     float64
Unnamed: 8     float64
Unnamed: 9     float64
Unnamed: 10    float64
Unnamed: 11    float64
Unnamed: 12    float64
Unnamed: 13    float64
dtype: object

##### Datatype of District, Month and Visitors column is 'object' and Year column has 'int64' datatype.

In [10]:
df.describe(include='object')

District  Month Visitors
count       6624   6624     6496
unique        70     22     2565
top     Adilabad  April        0
freq         240    552     2332

In [11]:
df.describe(include='number')

Unnamed: 0         Year  Unnamed: 3  Unnamed: 4  Unnamed: 5  \
count  6670.000000  6670.000000         0.0         0.0         0.0   
mean    180.061619  2018.928936         NaN         NaN         NaN   
std     109.766383     2.665551         NaN         NaN         NaN   
min       0.000000  2014.000000         NaN         NaN         NaN   
25%      83.000000  2017.000000         NaN         NaN         NaN   
50%     177.000000  2019.000000         NaN         NaN         NaN   
75%     275.000000  2021.000000         NaN         NaN         NaN   
max     395.000000  2023.000000         NaN         NaN         NaN   

       Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9  Unnamed: 10  \
count         0.0         0.0         0.0         0.0          0.0   
mean          NaN         NaN         NaN         NaN          NaN   
std           NaN         NaN         NaN         NaN          NaN   
min           NaN         NaN         NaN         NaN          NaN   
25%           NaN         NaN         NaN         NaN          NaN   
50%           NaN         NaN         NaN         NaN          NaN   
75%           NaN         NaN         NaN         NaN          NaN   
max           NaN         NaN         NaN         NaN          NaN   

       Unnamed: 11  Unnamed: 12  Unnamed: 13  
count          0.0          0.0          0.0  
mean           NaN          NaN          NaN  
std            NaN          NaN          NaN  
min            NaN          NaN          NaN  
25%            NaN          NaN          NaN  
50%            NaN          NaN          NaN  
75%            NaN          NaN          NaN  
max            NaN          NaN          NaN

In [12]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6670 entries, 0 to 6669
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   6670 non-null   int64  
 1   District     6624 non-null   object 
 2   Month        6624 non-null   object 
 3   Visitors     6496 non-null   object 
 4   Year         6670 non-null   int64  
 5   Unnamed: 3   0 non-null      float64
 6   Unnamed: 4   0 non-null      float64
 7   Unnamed: 5   0 non-null      float64
 8   Unnamed: 6   0 non-null      float64
 9   Unnamed: 7   0 non-null      float64
 10  Unnamed: 8   0 non-null      float64
 11  Unnamed: 9   0 non-null      float64
 12  Unnamed: 10  0 non-null      float64
 13  Unnamed: 11  0 non-null      float64
 14  Unnamed: 12  0 non-null      float64
 15  Unnamed: 13  0 non-null      float64
dtypes: float64(11), int64(2), object(3)
memory usage: 833.9+ KB


In [13]:
df['District'].unique() 

array(['Adilabad', 'Komaram Bheem Asifabad', 'Mancherial', 'Nirmal',
       'Hyderabad, Ranga Reddy, Medchal -Malkajigiri, & Vikarabad',
       'Karimnagar', 'Jagtial', 'Peddapalli', 'Rajarina Sircilla',
       'Khammam', 'Bhadradri Kothagudem', 'Mababubnagar',
       'Jogulamha Gadwal', 'Nagarkurnool', 'Wanaparthy', 'Narayanapet',
       'Medak', 'Sangareddy', 'Siddipet', 'Nalgonda & Suryapet',
       'Yadadri Bhongir', 'Nizamabad', 'Kamareddy', 'Hanumakonda',
       'Warangal (Rural)', 'Jangann', 'Jayashankar Bhoopalpally',
       'Mahabubabad', 'Mulugu', 'Hunmakonda',
       'Hyderabad, Ranga Reddy, Medchal & Vikarabad', 'Jangaon',
       'Jogulamba Gadwal', 'Mahbubnagar', 'Nalgonda', 'Narayanpet',
       'Peddapa li', 'Rajanna Sircilla', 'Siddip et', 'Suryapet',
       'Yadadri Bhongir •', 'Mahabubnagar', 'Warangal',
       'Bhadradri Kothagudem ', 'Hyderabad', 'Jagtial ', 'Jangaon ',
       'Jogulamba Gadwal ', 'Kamareddy ', 'Karimnagar ', 'Mahabubabad ',
       'Medak ', 'Medchal

In [14]:
df['Month'].unique() 

array(['January', 'February', 'March', 'April', 'May', 'June', 'July',
       'August', 'September', 'October', 'November', 'December',
       'February  ', 'March ', ' September', 'October   ', 'March  ',
       'May  ', 'October ', 'December  ', 'January ', 'May ', nan],
      dtype=object)

### Data Cleaning

In [15]:
# drop unnecessary columns
drop_columns = ['Unnamed: 3','Unnamed: 4','Unnamed: 5','Unnamed: 6','Unnamed: 7','Unnamed: 8',
                'Unnamed: 9','Unnamed: 10','Unnamed: 11','Unnamed: 12','Unnamed: 13']

df.drop(columns = drop_columns, axis=1, inplace=True)

In [16]:
# rename column
df.rename(columns={"Unnamed: 0": "Visitor's_ID"}, inplace=True)

In [17]:
# Remove leading and trailing spaces
df['District'] = df['District'].str.strip() 
df['Month'] = df['Month'].str.strip()

In [18]:
# Remove NaN values from the 'District' column
df['District'] = df['District'].dropna()

In [19]:
# Remove bullet point
df['District'] = df['District'].str.replace('•', '', regex=False)  

In [20]:
# replace '&' with ',' 
df['District'] = df['District'].str.replace('&', ',')

In [21]:
# Replace 'Siddip et' with 'Siddipet' in the 'District' column
df['District'] = df['District'].str.replace('Siddip et', 'Siddipet')

In [22]:
# Replace 'Siddip et' with 'Siddipet' in the 'District' column
df['District'] = df['District'].str.replace('Peddapa li', 'Peddapali') 

In [23]:
# Replace 'JogulambaGadwai' with 'Jogulamba Gadwal' in the 'District' column
df['District'] = df['District'].str.replace('JogulambaGadwai', 'Jogulamba Gadwal') 

In [24]:
# Replace 'Nalgonda & Suryapet' with 'Nalgonda , Suryapet' in the 'District' column
df['District'] = df['District'].str.replace('Nalgonda & Suryapet', 'Nalgonda , Suryapet')

In [25]:
# Replace 'JogulambaGadwai' with 'Jogulamba Gadwal' in the 'District' column
df['District'] = df['District'].str.replace('JogulambaGadwaI', 'Jogulamba Gadwal')

In [26]:
# Replace 'Jayashankarnhoopalpally' with 'Jayashankar Bhupalpally' in the 'District' column
df['District'] = df['District'].str.replace('Jayashankarnhoopalpally', 'Jayashankar Bhupalpally')
df['District'] = df['District'].str.replace('Jayashankar Bhoopalpally', 'Jayashankar Bhupalpally')

In [27]:
# Replace '& Vikarabad' with 'Vikarabad' in the 'District' column
df['District'] = df['District'].str.replace('Hyderabad, Ranga Reddy, Medchal -Malkajigiri, & Vikarabad', 
                                            'Hyderabad, Ranga Reddy, Medchal -Malkajigiri, Vikarabad')

In [28]:
# Replace 'jangann' with 'jangaon' in the 'District' column
df['District'] = df['District'].str.replace('Jangann', 'Jangaon')

In [29]:
# Replace 'mahabubnagar' with 'mahbubnagar' in the 'District' column
df['District'] = df['District'].str.replace('mahabubnagar', 'mahbubnagar')

In [30]:
# Replace 'BhadradriKothagudem' with 'Bhadradri Kothagudem' in the 'District' column
df['District'] = df['District'].str.replace('BhadradriKothagudem', 'Bhadradri Kothagudem')

In [31]:
# Replace 'Hunmakonda' with 'Hanumakonda' in the 'District' column
df['District'] = df['District'].str.replace('Hunmakonda', 'Hanumakonda')

In [32]:
# Replace 'Mababubnagar' with 'Mahbubnagar' in the 'District' column
df['District'] = df['District'].str.replace('Mababubnagar', 'Mahbubnagar')

In [33]:
# Replace 'KornaramBheentAsiTabad' with 'Kumram Bheem Asifabad' in the 'District' column
df['District'] = df['District'].str.replace('KornaramBheentAsiTabad', 'Kumram Bheem Asifabad')
df['District'] = df['District'].str.replace('KommramBheentAsifabad', 'Kumram Bheem Asifabad')

In [34]:
# Replace 'RajarinaSircilla' with 'Rajarina Sircilla' in the 'District' column
df['District'] = df['District'].str.replace('RajanuaSircilla', 'Rajanna Sircilla')
df['District'] = df['District'].str.replace('Rajarina Sircilla', 'Rajanna Sircilla')

In [35]:
# to replace the concatenated names with the correct district names
district_mapping = {'KommramBheentAsiTabad': 'Kumram Bheem Asifabad'}
df['District'] = df['District'].replace(district_mapping) 

In [36]:
district_mapping = {'YadadriBhongir': 'Yadadri Bhongir'}
df['District'] = df['District'].replace(district_mapping)

In [37]:
district_mapping = {'RajannaSircilla': 'Rajanna Sircilla'}
df['District'] = df['District'].replace(district_mapping)

In [38]:
df['District'].unique() 

array(['Adilabad', 'Komaram Bheem Asifabad', 'Mancherial', 'Nirmal',
       'Hyderabad, Ranga Reddy, Medchal -Malkajigiri, , Vikarabad',
       'Karimnagar', 'Jagtial', 'Peddapalli', 'Rajanna Sircilla',
       'Khammam', 'Bhadradri Kothagudem', 'Mahbubnagar',
       'Jogulamha Gadwal', 'Nagarkurnool', 'Wanaparthy', 'Narayanapet',
       'Medak', 'Sangareddy', 'Siddipet', 'Nalgonda , Suryapet',
       'Yadadri Bhongir', 'Nizamabad', 'Kamareddy', 'Hanumakonda',
       'Warangal (Rural)', 'Jangaon', 'Jayashankar Bhupalpally',
       'Mahabubabad', 'Mulugu',
       'Hyderabad, Ranga Reddy, Medchal , Vikarabad', 'Jogulamba Gadwal',
       'Nalgonda', 'Narayanpet', 'Peddapali', 'Suryapet',
       'Yadadri Bhongir ', 'Mahabubnagar', 'Warangal', 'Hyderabad',
       'Medchal', 'Ranga Reddy', 'Vikarabad', 'Warangal (Urban)',
       'Hyd , R R Dist.', 'Kumram Bheem Asifabad', nan], dtype=object)

In [39]:
df['Month'].nunique() 

12

In [40]:
# Remove rows with any null values
df = df.dropna() 

In [41]:
df.isnull().sum()

Visitor's_ID    0
District        0
Month           0
Visitors        0
Year            0
dtype: int64

In [42]:
null_percentage = (df.isnull().sum() / len(df)) * 100
null_percentage

Visitor's_ID    0.0
District        0.0
Month           0.0
Visitors        0.0
Year            0.0
dtype: float64

##### There are no nulls in the dataset.

In [43]:
# fill 0 inplace of null values
df['Visitors'] = pd.to_numeric(df['Visitors'], errors='coerce').fillna(0).astype('int64')

In [44]:
# Convert the 'Month' column to category
df['Month'] = df['Month'].astype('category') 

In [45]:
# sums the duplicate rows
df.duplicated().sum() 

360

##### There are 360 duplicate rows in the dataset.

In [46]:
duplicates = df[df.duplicated()]
duplicates.head(3) 

Visitor's_ID   District    Month  Visitors  Year
4034           100  Kamareddy      May         0  2021
4190           256  Nizamabad      May         0  2021
4462           180    Medchal  January         0  2020

In [47]:
# Group by 'District' and 'Month' and then check the uniqueness of 'Visitors'
partial_duplicates = df[df.duplicated(subset=['District', 'Month'], keep=False) & ~df.duplicated(subset=['District', 'Month', 'Visitors'], keep=False)]
partial_duplicates.head() 

Visitor's_ID                                           District     Month  \
48            48  Hyderabad, Ranga Reddy, Medchal -Malkajigiri, ...   January   
49            49  Hyderabad, Ranga Reddy, Medchal -Malkajigiri, ...  February   
50            50  Hyderabad, Ranga Reddy, Medchal -Malkajigiri, ...     March   
51            51  Hyderabad, Ranga Reddy, Medchal -Malkajigiri, ...     April   
52            52  Hyderabad, Ranga Reddy, Medchal -Malkajigiri, ...       May   

    Visitors  Year  
48     20226  2023  
49     18838  2023  
50     20599  2023  
51      5802  2023  
52      8674  2023

In [48]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
Index: 6496 entries, 0 to 6669
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   Visitor's_ID  6496 non-null   int64   
 1   District      6496 non-null   object  
 2   Month         6496 non-null   category
 3   Visitors      6496 non-null   int64   
 4   Year          6496 non-null   int64   
dtypes: category(1), int64(3), object(1)
memory usage: 260.5+ KB


In [49]:
df.to_csv('Cleaned_tourism_data.csv', index=False) 

In [50]:
import mysql.connector
import os

# List of CSV file and their corresponding table name
csv_file = [ ('Cleaned_tourism_data.csv', 'Tourism') ]

# Connect to the MySQL database
conn = mysql.connector.connect(
    host='127.0.0.1',
    user='root',
    password='Pandya@1939',
    database='tourism'
)
cursor = conn.cursor()

In [51]:
if conn.is_connected():
        print("Yes!")
cursor.execute("SELECT DATABASE();")
record = cursor.fetchone()
print("You're connected to database!")

Yes!
You're connected to database!


In [52]:
# Read the CSV file into a pandas DataFrame
csv_file_path = 'C:/ProgramData/MySQL/MySQL Server 8.0/Uploads/Cleaned_tourism_data.csv'
df = pd.read_csv(csv_file_path)
df.head(3) 

Visitor's_ID  District     Month  Visitors  Year
0             0  Adilabad   January         0  2023
1             1  Adilabad  February         0  2023
2             2  Adilabad     March         0  2023

In [53]:
# Replace NaN with None to handle SQL NULL
df = df.where(pd.notnull(df), None)

# Clean column names
df.columns = [col.replace(' ', '_').replace('-', '_').replace('.', '_') for col in df.columns]

# Define a function to map pandas dtypes to SQL data types
def get_sql_type(dtype):
    if pd.api.types.is_integer_dtype(dtype):
        return 'INT'
    elif pd.api.types.is_float_dtype(dtype):
        return 'FLOAT'
    elif pd.api.types.is_bool_dtype(dtype):
        return 'BOOLEAN'
    elif pd.api.types.is_datetime64_any_dtype(dtype):
        return 'DATETIME'
    else:
        return 'TEXT'

# Generate the CREATE TABLE statement with appropriate data types
table_name = 'tourism_data'
columns = ', '.join([f'`{col}` {get_sql_type(df[col].dtype)}' for col in df.columns])
create_table_query = f'CREATE TABLE IF NOT EXISTS `{table_name}` ({columns})'
cursor.execute(create_table_query)

# Insert DataFrame data into the MySQL table
for _, row in df.iterrows():
    # Convert row to tuple and handle NaN/None explicitly
    values = tuple(None if pd.isna(x) else x for x in row)
    sql = f"INSERT INTO `{table_name}` ({', '.join(['`' + col + '`' for col in df.columns])}) VALUES ({', '.join(['%s'] * len(row))})"
    cursor.execute(sql, values)

# Commit the transaction and close the connection
conn.commit()
cursor.close()
conn.close()

print("Data has been successfully loaded into the MySQL database.")

Data has been successfully loaded into the MySQL database.
